fontir/src/ir.rs

```diff
@@ -115,41 +116,122 @@ pub struct GlyphInstance {
     /// Advance height; if None, assumed to equal font's ascender - descende.
     pub height: Option<f64>,
     /// List of glyph contours.
-    pub contours: Vec<Contour>,
+    pub contours: Vec<BezPath>,
     /// List of glyph components.
     pub components: Vec<Component>,
 }
 
-/// A single glyph contour consisting of a list of points.
-pub type Contour = Vec<ContourPoint>;
-
-/// A single point in a glyph contour.
-#[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
-pub struct ContourPoint {
-    pub x: f64,
-    pub y: f64,
-    pub typ: PointType,
-}
-
-/// Possible types of a point in a glyph contour, following UFO GLIF semantics.
-#[derive(Serialize, Deserialize, Debug, Clone, PartialEq, Eq)]
-pub enum PointType {
-    Move,
-    Line,
-    OffCurve,
-    Curve,
-    QCurve,
-}
-
 /// A single glyph component, reference to another glyph.
 #[derive(Serialize, Deserialize, Debug, Clone, PartialEq)]
 pub struct Component {
     /// The name of the referenced glyph.
-    pub base: String,
+    pub base: GlyphName,
     /// Affine transformation to apply to the referenced glyph.
     pub transform: Affine,
 }
 
+/// Helps convert points-of-type to a bezier path.
+///
+/// Source formats tend to use streams of point-of-type. Curve manipulation is
+/// often easier on bezier path, so provide a mechanism to convert.
+#[derive(Debug)]
+pub struct GlyphPathBuilder<'a> {
+    glyph_name: &'a str,
+    offcurve: Vec<Point>,
+    path: BezPath,
+    last_move_to: Point,
+}
+
+impl<'a> GlyphPathBuilder<'a> {
+    pub fn new(glyph_name: &'a str) -> GlyphPathBuilder {
+        GlyphPathBuilder {
+            glyph_name,
+            offcurve: Vec::new(),
+            path: BezPath::new(),
+            // kurbo *requires* subpaths start with move so this isn't going to get used
+            last_move_to: Point::ZERO,
+        }
+    }
+
+    fn check_num_offcurve(&self, expected: impl Fn(usize) -> bool) -> Result<(), WorkError> {
+        if !expected(self.offcurve.len()) {
+            return Err(WorkError::InvalidSourceGlyph {
+                glyph_name: self.glyph_name.into(),
+                message: format!(
+                    "{} unexpected consecutive offcurve points {:?}",
+                    self.offcurve.len(),
+                    self.offcurve
+                ),
```

Comment:
An observation: seeing this makes me think it might make sense to have a distinct `GlyphConversionError` type, which can have variants representing various possible error conditions?

The thing that stands out here is that you're sort of using the `message` field as an error type, and the format string that you provide looks like it _really_ just wants to be the `Display` impl for your error. Instead, you could have a `TooManyOffCurves` error variant that cloned `self.offcurve`, and then you would produce this string in the `Display` impl.

Reply:
Hm. Yeah, probably right. Just being lazy. Will fix.